### ISSUE TITLE
TdmsFile.open() can use more memory than TdmsFile.read() for files with many segments and channels

### ISSUE THREAD

**Issue by Achilles1515** [adamreeve/npTDMS#203]
This is a little edge-casey but I ran into this issue recently in a "production" scenario -
Somehow or another (I was not in charge of this so I don't know all the details) our LabVIEW/VeriStand channel configuration was setup incorrectly and we ended up creating TDMS files that contained **~1600 channels** of data. These files are ~500 MB in size. And there are supposed to only be 13 channels, but a couple new ones that were added ended up being "duplicated" (in some type of manner) like so:

```
channel_name
channel_name #1
channel_name #2
...
channel_name #769
```
This problem caused a **huge** amount of memory to be consumed while using `TdmsFile.open()`. In the npTDMS library, this large memory usage occurs in the `TdmsReader._build_index()` method, these lines specifically:

```python
def _build_index(self):
    # ....
    segment_num_values = {
        path: np.zeros(num_segments, dtype=np.int64) for path in data_objects}
    segment_chunk_sizes = {
        path: np.zeros(num_segments, dtype=np.int64) for path in data_objects}

```

where `segment_num_values` eventually turns into:

```python
self._segment_channel_offsets = {
    path: np.cumsum(segment_count) for (path, segment_count) in
    segment_num_values.items()}
```

With the above TDMS file I described, `data_objects` has a length of 1598 and num_segments is 57839.

The resulting approximate memory usage of these three NumPy arrays (identical in size) are:
```
3 * (num_segments * sizeof(int64) * len(data_objects))
    = 3 * (57839 * 8 * 1598) = 2218241328 bytes
```

So something like 2.1 GB (and about 1.4 GB after a garbage collection run freeing `segment_num_values`), which is **huge** considering the total file size is only 500 MB.

Again, these TDMS files are setup incorrectly on our end, but are legitimate nonetheless. This huge memory usage can be rectified by decreasing the number of channels, but I am curious if it would be wise to add logic like the following:

* Before lazy-loading any channel data, perform the above segment index memory usage calculation for a file. At minimum, if the result exceeds the file size, then don't lazy-load channel data and instead just read the entire file in a similar manner to `TdmsFile.read()`.

(assuming no changes are made to how these indexes are being built)

Because, as the title of this issue states, `TdmsFile.open()` was using 3-4x more memory than simply reading the whole file upfront with `TdmsFile.read()`, on top of having slower data accessing time. So the logic stated in that bullet point is to avoid this lose-lose situation with using `TdmsFile.open()`, which directly contrasts the "memory efficiency" noted in the documentation:

>When you are working with large TDMS files or don't need to read all channel data, you can instead use `TdmsFile.open`. This is more memory efficient but accessing data can be slower

**Comment by adamreeve** [comment]
Thanks for the bug report, this does seem pretty edge-casey though as you say. I'm assuming in your scenario there might have only been a small number of segments with any data in per channel? I wonder if it would help to be a bit smarter about how the index is built, we could keep track of the first and last segments that have data for a channel, and then create arrays of length (last data segment - first data segment) per channel instead of the total number of segments, and then keep track of the segment offset.

If the data was more spread out across segments then this isn't so helpful, and in a scenario where there's only one data-point per channel per segment the index is still going to consume a lot more data than the actual channel data. So maybe your approach would be the most straightforward.

**Comment by HudsonMC16** [comment]
I have also observed this issue, but with much larger tdms files (2-5 GB). We have 600ish channels and a pretty long runtime. Probably not the most efficient way to store data, I know, but that's what we've got. I probably can't share the raw file, but I can share any details about the file which might be helpful in diagnosing this.

**Comment by adamreeve** [comment]
I had another look at this (apologies it's taken a while) and realised that in most cases there will be a lot of duplication in the `_segment_channel_offsets` arrays, so memory can be reduced by reusing arrays that are the same.

I've been working on improving the memory efficiency when using `TdmsFile.Open` in the `index_mem` branch and have made a number of optimisations:
* Add a per-segment index from object path to index in the objects array. These can be re-used across segments so shouldn't increase memory usage much in most cases, but the time taken for the initial `TdmsFile.Open` call will increase slightly (around 10% increase from my testing with a file where the segment objects alternate in each segment).
* This allows us to completely get rid of the `_segment_chunk_sizes` property, halving the index memory usage.
* De-duplicate the `_segment_channel_offsets` arrays as described above. In my test example this brings memory usage from around 100 MB to 1 MB, but this will depend a lot on the file structure.
* Build an index per-channel as required rather than all channels at once. So if you don't read data from all channels you don't pay the CPU and memory cost of building the index. And if you're working interactively you don't have a big wait to build all indexes at once on the first read.

**Comment by amylizzle** [comment]
I think this issue a little less edge-casey than it seems. I'm working with some data which has ~1000 groups per file, with each group having 16 channels, which are np.arrays of 3200x1.
Calling open and grabbing only the channels I need is eating 8Gb of RAM and is much slower than calling read which only uses 3Gb.

I might be able to provide a sample file if that would help with debugging, but I'd have to check with the data's owners. Let me know if that would be useful.

**Comment by adamreeve** [comment]
Hi @amylizzle, it shouldn't be necessary to provide an example, but are you able to test using the current code on the index_mem branch and see if that improves memory usage? I'd also be interested in whether you see much of an increase in time taken for the initial file open or time for the first channel read.

**Comment by amylizzle** [comment]
It's much faster to load using open() with the current commit - comparable with read(), but the memory usage is still significantly higher (4Gb on read(), 7Gb on open()). Before reading the channel data, open() only grabs 2Gb of RAM, which is still a little more than I'd expect for just indexing.

**Comment by amylizzle** [comment]
Oh yeah, the index_mem branch is hugely better. About 0.75Gb reserved after the call to open, and 1.6Gb after all the channels have been read once. It's reading a lot slower, but that's exactly what you'd expect when streaming from disk instead of RAM. 

Confirming index_mem branch fixes the issue for me. Thanks!

**Comment by adamreeve** [comment]
Thanks for testing @amylizzle, I've merged those changes to master so will close this as fixed now, although for files with a very irregular structure where the channel indexes aren't duplicated, using `TdmsFile.open` may still use a significant amount of memory.

**Comment by HudsonMC16** [comment]
@adamreeve Sorry I missed this discussion over the past few days, but looks like you got it fixed. I will test it on my end. What is your schedule for pushing to pypi? I can grab the repo for testing, but deployed stuff is pulling from pypi, so I'm just curious.

**Comment by adamreeve** [comment]
Hi @HudsonMC16, I'm aiming to make a new release within a week or so, I've got some other smaller performance improvements I'm testing too that I'd like to include.

**Comment by HudsonMC16** [comment]
Wonderful! Thanks, and thanks for the useful library!

**Comment by adamreeve** [comment]
This has now been released in version 1.2.0
